src/token.ts

```diff
@@ -1,19 +1,116 @@
 import { BaseChatModel } from '@langchain/core/language_models/chat_models';
-import { Token } from '@lumino/coreutils';
+import { ReadonlyPartialJSONObject, Token } from '@lumino/coreutils';
 import { ISignal } from '@lumino/signaling';
+import { JSONSchema7 } from 'json-schema';
 
 import { IBaseCompleter } from './llm-models';
 
+export interface IDict<T = any> {
+  [key: string]: T;
+}
+
+export interface IType<T> {
+  new (...args: any[]): T;
+}
+
+/**
+ * The provider interface.
+ */
 export interface IAIProvider {
+  /**
+   * The name of the provider.
+   */
   name: string;
-  completer: IBaseCompleter | null;
-  chatModel: BaseChatModel | null;
-  providerChanged: ISignal<IAIProvider, void>;
-  chatError: string;
-  completerError: string;
+  /**
+   * The chat model class to use.
+   */
+  chatModel?: IType<BaseChatModel>;
+  /**
+   * The completer class to use.
+   */
+  completer?: IType<IBaseCompleter>;
+  /**
+   * the settings schema for the provider.
+   */
+  settingsSchema?: any;
+  /**
+   * The instructions to be displayed in the settings, as helper to use the provider.
+   * A markdown renderer is used to render the instructions.
+   */
+  instructions?: string;
+  /**
+   * A function that extract the error message from the provider API error.
+   * Default to `(error) => error.message`.
+   */
+  errorMessage?: (error: any) => string;
+}
+
+/**
+ * The provider registry interface.
+ */
+export interface IAIProviderRegistry {
+  /**
+   * Get the list of provider names.
+   */
+  readonly providers: string[];
+  /**
+   * Add a new provider.
+   */
+  add(provider: IAIProvider): void;
+  /**
+   * Remove a provider.
+   */
+  remove(name: string): void;
```

Comment:
Wondering if we should expose a `remove` function?

Keeping registries "additive" would help prevent other extensions from removing existing providers.

Reply:
So we should also check if the provider already exist before setting it, to make it "additive" only.

Reply:
Right, so there is only one provider with the same name.

If an extension would like to replace an existing provider, it could disable the plugin providing it and add it in its own plugin.

Reply:
Done